pyomo/core/base/constraint.py

```diff
@@ -155,164 +151,17 @@ class _ConstraintData(ActiveComponentData):
         _active         A boolean that indicates whether this data is active
     """
 
-    __slots__ = ()
+    __slots__ = ('_body', '_lower', '_upper', '_expr')
```

Comment:
I see, thanks for the explanation. I'm all for removing the ABC layer, but if we are moving towards a duck-typing approach, do we need to make the data classes public? I.e., are `ctype` and `isindexed()` sufficient to do the kinds of checking that people need? I guess the intention for public data classes is for type-hinting? And potentially nicer errors than `AttributeError` if some random object is used somewhere instead of a constraint.

Reply:
That's a good question.  In the original world order, we convinced ourselves that a user would never need to directly interact with the component data classes (except through their public API) - which is why the class names were made "private".  With duck-typing, you could make a good argument that we could keep them private.  That said, I think the strongest argument is for documentation (littering the type hints with underscored classes - that are not picked up automatically by things like autodoc - was "ugly"). 

There is also the point that just because we *want* to support duck-typing, doesn't mean that we won't allow inheritance.  For example, IDAES, et al., use inheritance extensively.  And in the component object model, the thing you inherit from is the *data* class (which up to now was "private").